### PR TITLE
Add getCharCode method to Buri string type

### PR DESCRIPTION
**Pull request by aaronstanek**

### CODE REVIEW

File: packages/std/prelude/string.js [hunk 1]
```diff
@@ -6,4 +8,11 @@ Object.assign(String.prototype, {
     size() {
         return this.length
     },
+
+    codePointAt(index) {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: I'm not sure if this name is self explanatory. Would `charAt` be appropriate? Or even `get` to make the API consistent with lists?

File: rust/type_checker/types/src/default_types.rs [hunk 2]
```diff
@@ -37,11 +37,18 @@ pub fn create_string_default_methods(
...
-            arguments: vec![],
-            return_type: INT_TYPE_ID,
-        }],
+        vec![
+            Method {
+                name: "size",
+                arguments: vec![],
+                return_type: INT_TYPE_ID,
+            },
+            Method {
```
Comment (Nick-Mazuk) [inline rust/type_checker/types/src/default_types.rs:46]: Yes, add this here. However note that it'll cause a lot of tests in the [parsed_expression_to_generic_expression.rs](https://github.com/cambiata-team/buri/blob/main/rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs) because it's adding new constraints. So any test that hardcodes the number of constraints that involves strings will need to be updated as well.

File: tests/js/valid/strings/codePointAt.test.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+it("\"hello\":codePointAt(0) == #some(104)", () => {
+    expect(BcodeOfLowercaseH).toEqual(tag("some", 104))
+})
+
+it("\"hello\":codePointAt(1) == #some(101)", () => {
+    expect(BcodeOfLowercaseE).toEqual(tag("some", 101))
+})
+
+it("\"hello\":codePointAt(-1) == #none", () => {
```
Comment (Nick-Mazuk) [inline tests/js/valid/strings/codePointAt.test.js]: It may be nice for the API to wrap around, so `-1` retrieves the last element, `-2` the second-to-last, etc.

JavaScript's `at` method is a good example of this. And I believe Python does something similar.

If we do this, then `|index| >= self.length` would return `#none`.

**Review by Nick-Mazuk — CHANGES_REQUESTED** [review]
Just want to ensure any methods we add are well thought-out since they will be difficult (though not impossible) to change later.

File: packages/std/prelude/string.js (hunk 1 above)
Comment (aaronstanek) [reply packages/std/prelude/string.js]: I agree that the permanence of the names of built-in functions warrants a higher degree of thought and scrutiny. For this reason, I encourage the most challenging feedback you can provide.

Just so we are on the same page, this method returns the nth Unicode scalar value in a string. It's return type is `Int`. The behavior of this method is distinct from getting the nth character in a string, which in Buri would have a return type of `Str`.

The decision criteria which resulted in me selecting `codePointAt` were the following:
1. The name must be concise and descriptive.
2. The name must not be confused for a method that returns the nth character of a string.
3. The name must not be confused for a method that converts a string to a list of Unicode scalar values.
4. If possible, the name should be similar to equivalent methods in other languages.

I think that your suggestions of `charAt` and `get` would be really good for a method that returns the nth character of a string. However, these suggestions don't work too well with my second criterion. I like the idea of having API consistency between lists and strings, and I think it would be a great idea to have a `get` method on strings that returns the nth character.

Aside from that, the fact that you find `codePointAt` less than self-explanatory is a pretty big blow to it's candidacy under my first criterion. I'll probably go back to the drawing board here.

The reasons why I liked `codePointAt` were the following:
- It was relatively short
- I found it self-descriptive
- I figured that by using the term "code point" it was unlikely to be confused for a method that returns a character
- I figured that by using the term "at" it was unlikely to be confused for a method that converts a string to a list of Unicode scalar values
- Both Java and JavaScript call their equivalent to this method `codePointAt`

What I didn't like about the name `codePointAt` is that it implies that it returns a Unicode code point, when in actuality it returns a Unicode scalar value. The distinction between these two terms is small, but it may be important in some niche edge cases.

File: rust/type_checker/types/src/default_types.rs (hunk 2 above)
Comment (aaronstanek) [reply rust/type_checker/types/src/default_types.rs:46]: Thanks for the heads-up.

File: tests/js/valid/strings/codePointAt.test.js (hunk 3 above)
Comment (aaronstanek) [reply tests/js/valid/strings/codePointAt.test.js]: I agree, this is a good behavior.

File: packages/std/prelude/string.js (hunk 1 above)
Comment (aaronstanek) [reply packages/std/prelude/string.js]: Something like `scalarValueAt` is really specific, but it's unlikely to be understood by people who aren't familiar with Unicode terminology.

I am also considering:
- `codeAt`
- `charCodeAt`
- `valueAt`
- `charValueAt`
- `unicodeAt`

File: tests/js/valid/strings/codePointAt.test.js (hunk 3 above)
Comment (aaronstanek) [reply tests/js/valid/strings/codePointAt.test.js]: Implemented.

File: rust/type_checker/types/src/default_types.rs (hunk 2 above)
Comment (aaronstanek) [reply rust/type_checker/types/src/default_types.rs:46]: Fixed.

File: packages/std/prelude/string.js (hunk 1 above)
Comment (Nick-Mazuk) [reply packages/std/prelude/string.js]: Oh, I misinterpreted the purpose of this function.

Of these, I'd suggest `getCharCode`. `CarCode` because that's what JS calls it. `get` to make it similar to lists.

So…

- Str
  - get() – get's the nth character
  - getCharCode() - get's the nth character code
- List
  - get() - get's the nth item

Comment (aaronstanek) [reply packages/std/prelude/string.js]: I like this.

Comment (aaronstanek) [reply packages/std/prelude/string.js]: Name changed.

File: packages/std/prelude/string.js [hunk 4]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
     size() {
         return this.length
     },
+
+    /// Assume that the string is valid UTF-16.
+    /// Return the offset of the nth character in the string.
+    /// If n is negative, return the offset of the character that is nth from the end.
+    /// If the nth character is composed of a surrogate pair, return the offset of the high surrogate.
+    /// If n is out of bounds, return -1.
+    jsIndexForNthCharacter(n) {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: Could we move this out of the class? That way if someone does add the trait method `jsIndexForNthCharacter` to a `Str`, this won't be overwritten. It'll also make things (slightly) faster because it's one fewer method lookup.

File: packages/std/prelude/string.js [hunk 5]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
...
+        // ((this.charCodeAt(jsIndex) & 0xF800) === 0xD800) is true if and only if the code point at jsIndex is a UTF-16 surrogate.
+        if (n >= 0) {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: Both branches are almost exactly identical. Consider refactoring this to remove the if statement.

```js
let parity = n >= this.length ? 1 : -1
// ...
characterIndex += parity
jsIndex += (1 + ((this.charCodeAt(jsIndex) & 0xF800) === 0xD800)) * parity
```

File: packages/std/prelude/string.js [hunk 6]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
...
+            if (n >= this.length) {
+                return -1
+            }
+            let characterIndex = 0;
+            let jsIndex = 0;
+            while (true) {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: `while (true)` is almost always a code smell. Consider refactoring this:

```js
while (Math.abs(jsIndex) < this.length) {
    // ...
}
```

File: packages/std/prelude/string.js [hunk 7]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
...
         return this.length
     },
+
+    /// Assume that the string is valid UTF-16.
+    /// Return the offset of the nth character in the string.
+    /// If n is negative, return the offset of the character that is nth from the end.
+    /// If the nth character is composed of a surrogate pair, return the offset of the high surrogate.
+    /// If n is out of bounds, return -1.
+    jsIndexForNthCharacter(n) {
+        // ((this.charCodeAt(jsIndex) & 0xF800) === 0xD800) is true if and only if the code point at jsIndex is a UTF-16 surrogate.
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: Instead of explaining this in a comment, consider moving this into it's own function. That removes duplication and allows the code itself to explain itself.

```js
// this should probably even go outside the class
const isUtf16Surrogate = (charCode) => (charCode & 0xF800) === 0xD800

// when it's used
isUtf16Surrogate(this.charCodeAt(jsIndex))
```

File: packages/std/prelude/string.js [hunk 8]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
...
+            while (true) {
+                if (jsIndex >= this.length) {
+                    return -1
+                }
+                else if (characterIndex === n) {
+                    return jsIndex
+                }
+                else {
+                    characterIndex += 1
+                    jsIndex += 1 + ((this.charCodeAt(jsIndex) & 0xF800) === 0xD800)
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: It's generally not a good idea to rely on type coercion since it makes code harder to read (especially when it's implicit coercion like it is here).

File: packages/std/prelude/string.js [hunk 9]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
...
+            }
+        }
+    },
+
+    getCharCode(n) {
+        let jsIndex = this.jsIndexForNthCharacter(n)
+        if (jsIndex === -1) {
+            return ["none"]
+        }
+        else {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: Nit: else should be on the same line as the previous curly brace: `} else {`

File: packages/std/prelude/string.js [hunk 10]
```diff
@@ -6,4 +6,61 @@ Object.assign(String.prototype, {
     size() {
         return this.length
     },
+
+    /// Assume that the string is valid UTF-16.
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js]: If you want to add doc comments to JS code, use JSDoc comments. This is not only the industry standard, but it's also how you get intellisense in VS Code.

```js
/**
 * Assume the string is valid UTF-16.
 * ...
 */
jsIndexForNthCharacter(n) {
```

**Review by Nick-Mazuk — CHANGES_REQUESTED** [review]

File: packages/std/prelude/string.js (hunk 9 above)
Comment (Nick-Mazuk) [reply packages/std/prelude/string.js]: Hmm… looks like this is still unresolved.

File: packages/std/prelude/string.js [hunk 11]
```diff
@@ -1,3 +1,42 @@
+const isUtf16SurrogateAtIndex = (string, jsIndex) => (string.charCodeAt(jsIndex) & 0xF800) === 0xD800
+
+/**
+ * Assume that the string is valid UTF-16.
+ * Return the offset of the nth character in the string.
+ * If n is negative, return the offset of the character that is nth from the end.
+ * If the nth character is composed of a surrogate pair, return the offset of the high surrogate.
+ * If n is out of bounds, return -1.
+ */
+const jsIndexForNthCharacter = (string, n) => {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js:10]: Definitely much better and easier to read! Two things, though:

- Use ternary operators to eliminate uninitialized values.
- Remove else blocks after return statements to remove unnecessary nesting.

```js
let step = n >= 0 ? 1 : -1
let characterIndex = n >= 0 ? 0 : 1
let jsIndex = n >= 0 ? 0 : string.length - 1

while (jsIndex >= 0 && jsIndex < string.length) {
    if (characterIndex === n) {
        if (step === -1 && isUtf16SurrogateAtIndex(string, jsIndex)) {
            return jsIndex - 1
        }
        return jsIndex
    }

    characterIndex += step
    if (isUtf16SurrogateAtIndex(string, jsIndex)) {
        jsIndex += step * 2
    } else {
        jsIndex += step
    }
}

return -1
```

**Review by Nick-Mazuk — CHANGES_REQUESTED** [review]
After this, LGTM.

File: packages/std/prelude/string.js [hunk 12]
```diff
@@ -1,3 +1,36 @@
...
+    let characterIndex = n >= 0 ? 0 : -1
+    let jsIndex = n >= 0 ? 0 : string.length - 1
+
+    while (jsIndex >= 0 && jsIndex < string.length) {
+        if (characterIndex === n) {
+            if (step === -1 && isUtf16SurrogateAtIndex(string, jsIndex)) {
+                return jsIndex - 1
+            }
+            return jsIndex
+        } else {
```
Comment (Nick-Mazuk) [inline packages/std/prelude/string.js:21]: You still don't need this else here since all paths of the if statement `return`.

**Review by Nick-Mazuk — APPROVED** [review]
